cli/replicasets/topology_replicasets.go

```diff
@@ -0,0 +1,333 @@
+		"vshard_group": &replicaset.VshardGroup,
+	}
+
+	for key, valuePtr := range stringFieldsMap {
+		if err := getStringValueFromMap(replicasetMap, key, valuePtr); err != nil {
+			return nil, fmt.Errorf("Failed to get string fields: %s", err)
+		}
+	}
+
+	floatFieldsMap := map[string]*float64{
+		"weight": &replicaset.Weight,
+	}
+
+	for key, valuePtr := range floatFieldsMap {
+		if err := getFloatValueFromMap(replicasetMap, key, valuePtr); err != nil {
+			return nil, fmt.Errorf("Failed to get int fields: %s", err)
+		}
+	}
+
+	boolFieldsMap := map[string]*bool{
+		"all_rw": &replicaset.AllRW,
+	}
+
+	for key, valuePtr := range boolFieldsMap {
+		if err := getBoolValueFromMap(replicasetMap, key, valuePtr); err != nil {
+			return nil, fmt.Errorf("Failed to get bool fields: %s", err)
+		}
+	}
+
+	stringSliceFieldsMap := map[string]*[]string{
+		"roles": &replicaset.Roles,
+	}
+
+	for key, valuePtr := range stringSliceFieldsMap {
+		if err := getStringSliceValueFromMap(replicasetMap, key, valuePtr); err != nil {
+			return nil, fmt.Errorf("Failed to get string array fields: %s", err)
+		}
+	}
+
+	if err := getTopologyInstancesFromMap(replicasetMap, "instances", &replicaset.Instances); err != nil {
+		return nil, fmt.Errorf("Failed to get string array fields: %s", err)
+	}
+
+	return &replicaset, nil
+}
+
+func getTopologyInstancesFromMap(m map[string]interface{}, key string, valuePtr *TopologyInstances) error {
+	instancesRaw, found := m[key]
+	if !found {
+		return nil
+	}
+
+	instancesRawSlice, err := convertToSlice(instancesRaw)
+	if err != nil {
+		return fmt.Errorf("%q should be an array, got %#v", key, instancesRaw)
+	}
+
+	topologyInstances := make(TopologyInstances, len(instancesRawSlice))
+
+	for i, instanceRaw := range instancesRawSlice {
+		instanceRawMap, err := convertToMapWithStringKeys(instanceRaw)
+		if err != nil {
+			return fmt.Errorf("All elements of %q should be a map, got %#v", key, instancesRaw)
+		}
+
+		topologyInstance := TopologyInstance{}
+
+		stringFieldsMap := map[string]*string{
+			"alias": &topologyInstance.Alias,
+			"uuid":  &topologyInstance.UUID,
+		}
+
+		for key, valuePtr := range stringFieldsMap {
+			if err := getStringValueFromMap(instanceRawMap, key, valuePtr); err != nil {
+				return fmt.Errorf("Replicaset received in wrong format: %s", err)
+			}
+		}
+
+		topologyInstances[i] = &topologyInstance
+	}
+
+	*valuePtr = topologyInstances
+
+	return nil
+}
+
+func getStringValueFromMap(m map[string]interface{}, key string, valuePtr *string) error {
+	valueRaw, found := m[key]
+	if !found {
+		return nil
+	}
+
+	value, ok := valueRaw.(string)
+	if !ok {
+		return fmt.Errorf("%q value should be string, found %#v", key, valueRaw)
+	}
+
+	*valuePtr = value
+
+	return nil
+}
+
+func getFloatValueFromMap(m map[string]interface{}, key string, valuePtr *float64) error {
+	valueRaw, found := m[key]
+	if !found {
+		return nil
+	}
+
+	var value float64
+
+	switch i := valueRaw.(type) {
```

Comment:
I think it could be rewrite in more laconic way with reflect.

All int values you can cast with https://golang.org/pkg/reflect/#Value.SetInt and then to float. All float values you can convent with https://golang.org/pkg/reflect/#Value.SetFloat

Also please revise again is it really needed. E.g. json always decode values to float https://golang.org/pkg/encoding/json/#Token (`float64, for JSON numbers`)